File: HubWebApp/Internal/XTI_Hub/XtiVersion.cs
using XTI_App.Abstractions;
using XTI_Hub.Abstractions;
using XTI_HubDB.Entities;

namespace XTI_Hub;

public sealed class XtiVersion
{
    private readonly HubFactory factory;
    private readonly XtiVersionEntity record;

    internal XtiVersion(HubFactory factory, XtiVersionEntity record)
    {
        this.factory = factory;
        this.record = record ?? new XtiVersionEntity();
        ID = this.record.ID;
    }

    public int ID { get; }

    public AppVersionKey Key() => AppVersionKey.Parse(record.VersionKey);

    public Task Publishing() => factory.Versions.Publishing(record);

    public Task Published() => factory.Versions.Published(record);

    internal AppVersion App(App app) => new AppVersion(factory, app, this);

    public XtiVersionModel ToModel() => new XtiVersionModel
    {
        ID = ID,
        VersionName = new AppVersionName(record.VersionName),
        VersionKey = Key(),
        VersionNumber = new AppVersionNumber(record.Major, record.Minor, record.Patch),
        VersionType = AppVersionType.Values.Value(record.Type),
        Status = AppVersionStatus.Values.Value(record.Status),
        TimeAdded = record.TimeAdded
    };

    public override string ToString() => $"{nameof(XtiVersion)} {ID}: {Key()}";
}

File: HubWebApp/Apps/HubWebApp/Scripts/Lib/XtiVersion.ts
import { DateTimeOffset } from "@jasonbenfield/sharedwebapp/DateTimeOffset";
import { AppVersionKey } from "./AppVersionKey";
import { AppVersionName } from "./AppVersionName";
import { AppVersionNumber } from "./AppVersionNumber";
import { AppVersionStatus } from "./Http/AppVersionStatus";
import { AppVersionType } from "./Http/AppVersionType";

export class XtiVersion {
    readonly id: number;
    readonly versionName: AppVersionName;
    readonly versionKey: AppVersionKey;
    readonly versionNumber: AppVersionNumber;
    readonly versionType: AppVersionType;
    readonly status: AppVersionStatus;
    readonly timeAdded: DateTimeOffset;

    constructor(source?: IXtiVersionModel) {
        this.id = source?.ID || 0;
        this.versionName = new AppVersionName(source?.VersionName);
        this.versionKey = new AppVersionKey(source?.VersionKey);
        this.versionNumber = new AppVersionNumber(source?.VersionNumber);
        this.versionType = AppVersionType.values.value(source?.VersionType.Value || 0);
        this.status = AppVersionStatus.values.value(source?.Status.Value || 0);
        this.timeAdded = source?.TimeAdded || DateTimeOffset.max();
    }
}